Kernel/Devices/VMWareSVGADevice.cpp

```diff
@@ -0,0 +1,269 @@
+    m_fifo_addr = m_fifo_mapping->vaddr();
+    // Clear framebuffer memory
+    OwnPtr<Region> clear_map = MM.allocate_kernel_region(m_framebuffer_address_physical, page_round_up(PCI::get_BAR_space_size(address, 1)), "VMWareSVGA FB Test", Region::Access::Write, Region::Cacheable::No);
+    memset(clear_map->vaddr().as_ptr(), 0x00, PCI::get_BAR_space_size(address, 1));
+    clear_map->unmap();
+    VERIFY(m_io_base);
+    VERIFY(m_framebuffer_address_physical.as_ptr());
+    VERIFY(m_fifo_addr.as_ptr());
+
+    // Determine SVGA Version
+    // The way this is done is by writing the highest version we support (2) into a device support register,
+    // and checking if the value has been changed.
+    // If it hasn't, then we found the highest version that both we and the "card" support.
+    // If it has, decrement by one and repeat.
+    m_device_version = make_version_id(2);
+    set_register(IORegister::ID, m_device_version);
+    while (get_register(IORegister::ID) != m_device_version && m_device_version >= make_version_id(0)) {
+        m_device_version--;
+        set_register(IORegister::ID, m_device_version);
+    }
+
+    VERIFY(m_device_version >= make_version_id(0));
+
+    // We want to read the size of the Fifo memory
+    m_fifo_size = get_register(IORegister::MemSize);
+
+    // The reference driver checks some sizes here, but Qemu actually makes its buffers for
+    // the fifo smaller then VMWare does, so they fail. The fifo still works, however, and the
+    // numbers are accurate.
+
+    // Check if the card is new enough and get the capabilities register
+    if (m_device_version >= make_version_id(1)) {
+        m_device_capabilities = get_register(IORegister::Capabilities);
+    }
+
+    // Check if interrupt masks are supported
+    if (get_capability(Capabilities::IRQMask)) {
+        // Get the card interrupt line
+        m_interrupt_line = PCI::get_interrupt_line(address);
+        change_irq_number(m_interrupt_line);
+        // Mask all interrupts
+        set_register(IORegister::IRQMask, 0xFF);
+        // Clear all interrupts on the card
+        IO::out32(m_io_base + static_cast<u32>(Ports::IRQStatus), 0xFF);
+        // Enable the interrupt handler
+        enable_irq();
+    }
+
+    dbgln("VMWareSVGA: handshake complete, version {}, max res: {}x{}, irq: {}", m_device_version & 0xFF, get_register(IORegister::MaxWidth), get_register(IORegister::MaxHeight), m_interrupt_line);
+
+    // FIXME: Remove this when we get signals for updates from WindowServer
+    RefPtr<Thread> update_thread;
+    Process::create_kernel_process(update_thread, "VMWareSVGAThread", [&] {
+        dbgln_if(VMWARESVGA_DEBUG, "VMWareSVGA: Thread is running");
+        for (;;) {
+            force_vblank();
+            // We choose a time slightly smaller then 60Hz as
+            // we will never be perfectly in sync with the monitor
+            (void)Thread::current()->sleep(Time::from_milliseconds(15));
+        }
```

Comment:
Hmm.. so this creates a kernel thread that sleeps for 15ms, sends an update command to the VM, and then burns CPU for 1.666ms.

That seems like a very unfortunate design. Is there no way to get an IRQ from the graphics card?

Reply:
There isn't. The only three IRQs are when a fence in the Fifo passes, when a specific fence in the Fifo passes, and when progress was made in the Fifo.

Reply:
Forgot to say this, but i doubt force_vblank() burns for 1.666ms. It sends a update screen command to the card, and then clears the fifo.